docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+
+## Module Layout
+
+Modules form the basic units of imported code in Rust, so identifying where to split functionality and what to call modules is essential. I suspect this is the right level to divide things like collation vs segmentation etc. but it's also far too early to make hard recommendations about this, so I proposed to leave this issue open for the time being (moving code between modules later should be relatively easy).
+
+**Open Question**: Should we have code in the top-level module of a Crate or always use named sub-modules? This probably ties into what Crates we intend to publish and the granularity of the APIs.
+
+## Shared Internal Code
+
+Rust supports having non-public sub-modules within a Crate, and only the top-level "implicit" module is exposed by default (other modules can be exported by the top level module). See also [File hierarchy - Rust By Example](https://doc.rust-lang.org/rust-by-example/mod/split.html).
+
+### Use Private Modules :: required
+
+While local private functions can just be defined in any module, shared internal functions should always be implemented in a non-exported sub-module. This avoids any misleading dependencies between "public" modules and should help ensure clean code dependencies. Code in internal modules should still be tested and documented thoroughly. It is perfectly expected that we would have multiple internal modules for different aspects of ICU4X.
+
+# Structs and Traits
+
+For ICU4X we should be looking to keep code consistent and very unsurprising, especially for layout and behavior of data structures.
+
+## Private vs Public
+
+Rust offers a compelling model for encapsulating implementation details based on a hierarchical view of privacy; see [Visibility and Privacy - The Rust Reference](https://doc.rust-lang.org/reference/visibility-and-privacy.html). This should allow ICU4X to encapsulate all of its implementation details neatly from applications using the library and avoid too many cases of [Hyrum's Law](https://www.hyrumslaw.com/).
+
+### Minimize user-visible structs :: required
+
+While this sounds a bit obvious, I think it's important to stress that with Rust's privacy model, there is never any need to have an internal type be user-visible, and once a type is user-visible outside a Crate, it's published and should not be changed.
+
+The only place this might be an issue is if we decide to deliver ICU4X as multiple Crates. If we do that then we just added a requirement that our inter-Crate APIs be public (there's no privileged sharing outside Crates). This is one reason I propose we deliver ICU4X as a single Crate with multiple modules.
+
+### No public fields :: suggested
+
+Unless there is a compelling use case, it seems very sensible to never use public fields in structs. This allows for clean encapsulation with essentially no downside for ICU4X. This post sums things up nicely, which I am summarizing below:
```

Comment:
Overall the model I'm familiar with is that you have two kinds of types:

 - bag of stuff
 - "objects"

"bag of stuff" types, eg config structs, are basically a convenient wrapper around multiple types. These should have all pub fields

"object" types: these have semantics and should have no pub fields.

Reply:
oh, this is covered below

Reply:
Yeah, I think this is still TBD